README.md
# diypowerflow
DIY Powerflow

This a DIY version of the powerflow screen on the Tesla powerwall app.

It gets its values from MQTT topics and uses websockets for pushing the values into the client interface.

It includes a solar kwh prediction mechanism which allows for flexible solar panels configuration. It bases on Yr.no and darksky.net API forecasts.

I currently use MPP PCM60x charger (Solar), [Batrium BMS](https://www.batrium.com/) with a Shunt (Powerwall) and two SDM120 meters (House & Grid),
and a raspberry PI to get the values from all those components and post them to MQTT.

I will include more documentation (maybe some videos) explaining better how to do this exactly for people who need it.

If you have the same or similar components (like other MPP inverters) I suggest you check out the ISO for the raspberry that
Daniel Römer from [DIY Tech and Repairs](http://diytechandrepairs.nu/raspberry-solar/) has prepared,
since it includes already a lot of scripts which you can use and adapt for your own specific situation.

## How to run it
It requires a running MQTT server from which to get its power values. 
Also it uses influx for the kWh values and powerwall SOC.
So make sure you have both installed and running somewhere in your network.

Open project.js and configure it for your usecase.
then run
```
npm install
```
to ensure all dependecies are available

and then run
```
node project.js
```
in order for solar prediction mechanism to use darksky.net, you have to get an API key from https://darksky.net/dev

## TODO/ROADMAP:
* use grafana graphs for detailed information on each component
* show detailed house consumption from sonoffs
* improve dots animations, to better reflect current flowing
* make additional energy sources, like generator and wind turbines.
* ...? what ideas/suggestions do you have? join in!




CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0]
- now a node.js app with its own http server, no need for apache or external server
- removed emoncms support, since it became paid service
- uses mqtt + websockets for pushing power values, instead of pulling every x seconds
- kWh and SOC values are now read from influx db instead of emoncms
- added solar kWh prediction based on clouds forecasts (yr.no & darksky.net)and sun position calculation

## [1.0.0]
- initial commit with emoncms feed support


proxy.php
<?php

/**
 * AJAX Cross Domain (PHP) Proxy 0.8
 * Copyright (C) 2016 Iacovos Constantinou (https://github.com/softius)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Enables or disables filtering for cross domain requests.
 * Recommended value: true
 */
define('CSAJAX_FILTERS', true);

/**
 * If set to true, $valid_requests should hold only domains i.e. a.example.com, b.example.com, usethisdomain.com
 * If set to false, $valid_requests should hold the whole URL ( without the parameters ) i.e. http://example.com/this/is/long/url/
 * Recommended value: false (for security reasons - do not forget that anyone can access your proxy)
 */
define('CSAJAX_FILTER_DOMAIN', true);

/**
 * Set debugging to true to receive additional messages - really helpful on development
 */
define('CSAJAX_DEBUG', false);

/**
 * A set of valid cross domain requests
 */
$valid_requests = array(
    'emoncms.org'
);

/**
 * Set extra multiple options for cURL
 * Could be used to define CURLOPT_SSL_VERIFYPEER & CURLOPT_SSL_VERIFYHOST for HTTPS
 * Also to overwrite any other options without changing the code
 * See http://php.net/manual/en/function.curl-setopt-array.php
 */
$curl_options = array(
    // CURLOPT_SSL_VERIFYPEER => false,
    // CURLOPT_SSL_VERIFYHOST => 2,
);

/* * * STOP EDITING HERE UNLESS YOU KNOW WHAT YOU ARE DOING * * */

// identify request headers
$request_headers = array( );
foreach ($_SERVER as $key => $value) {
    if (strpos($key, 'HTTP_') === 0  ||  strpos($key, 'CONTENT_') === 0) {
        $headername = str_replace('_', ' ', str_replace('HTTP_', '', $key));
        $headername = str_replace(' ', '-', ucwords(strtolower($headername)));
        if (!in_array($headername, array( 'Host', 'X-Proxy-Url' ))) {
            $request_headers[] = "$headername: $value";
        }
    }
}

// identify request method, url and params
$request_method = $_SERVER['REQUEST_METHOD'];
if ('GET' == $request_method) {
    $request_params = $_GET;
} elseif ('POST' == $request_method) {
    $request_params = $_POST;
    if (empty($request_params)) {
        $data = file_get_contents('php://input');
        if (!empty($data)) {
            $request_params = $data;
        }
    }
} elseif ('PUT' == $request_method || 'DELETE' == $request_method) {
    $request_params = file_get_contents('php://input');
} else {
    $request_params = null;
}

// Get URL from `csurl` in GET or POST data, before falling back to X-Proxy-URL header.
if (isset($_REQUEST['csurl'])) {
    $request_url = urldecode($_REQUEST['csurl']);
} elseif (isset($_SERVER['HTTP_X_PROXY_URL'])) {
    $request_url = urldecode($_SERVER['HTTP_X_PROXY_URL']);
} else {
    header($_SERVER['SERVER_PROTOCOL'] . ' 404 Not Found');
    header('Status: 404 Not Found');
    $_SERVER['REDIRECT_STATUS'] = 404;
    exit;
}

$p_request_url = parse_url($request_url);

// csurl may exist in GET request methods
if (is_array($request_params) && array_key_exists('csurl', $request_params)) {
    unset($request_params['csurl']);
}

// ignore requests for proxy :)
if (preg_match('!' . $_SERVER['SCRIPT_NAME'] . '!', $request_url) || empty($request_url) || count($p_request_url) == 1) {
    csajax_debug_message('Invalid request - make sure that csurl variable is not empty');
    exit;
}

// check against valid requests
if (CSAJAX_FILTERS) {
    $parsed = $p_request_url;
    if (CSAJAX_FILTER_DOMAIN) {
        if (!in_array($parsed['host'], $valid_requests)) {
            csajax_debug_message('Invalid domain - ' . $parsed['host'] . ' does not included in valid requests');
            exit;
        }
    } else {
        $check_url = isset($parsed['scheme']) ? $parsed['scheme'] . '://' : '';
        $check_url .= isset($parsed['user']) ? $parsed['user'] . ($parsed['pass'] ? ':' . $parsed['pass'] : '') . '@' : '';
        $check_url .= isset($parsed['host']) ? $parsed['host'] : '';
        $check_url .= isset($parsed['port']) ? ':' . $parsed['port'] : '';
        $check_url .= isset($parsed['path']) ? $parsed['path'] : '';
        if (!in_array($check_url, $valid_requests)) {
            csajax_debug_message('Invalid domain - ' . $request_url . ' does not included in valid requests');
            exit;
        }
    }
}

// append query string for GET requests
if ($request_method == 'GET' && count($request_params) > 0 && (!array_key_exists('query', $p_request_url) || empty($p_request_url['query']))) {
    $request_url .= '?' . http_build_query($request_params);
}

// let the request begin
$ch = curl_init($request_url);
curl_setopt($ch, CURLOPT_HTTPHEADER, $request_headers);   // (re-)send headers
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);     // return response
curl_setopt($ch, CURLOPT_HEADER, true);       // enabled response headers
// add data for POST, PUT or DELETE requests
if ('POST' == $request_method) {
    $post_data = is_array($request_params) ? http_build_query($request_params) : $request_params;
    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_POSTFIELDS,  $post_data);
} elseif ('PUT' == $request_method || 'DELETE' == $request_method) {
    curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $request_method);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $request_params);
}

// Set multiple options for curl according to configuration
if (is_array($curl_options) && 0 <= count($curl_options)) {
    curl_setopt_array($ch, $curl_options);
}

// retrieve response (headers and content)
$response = curl_exec($ch);
curl_close($ch);

// split response to header and content
list($response_headers, $response_content) = preg_split('/(\r\n){2}/', $response, 2);

// (re-)send the headers
$response_headers = preg_split('/(\r\n){1}/', $response_headers);
foreach ($response_headers as $key => $response_header) {
    // Rewrite the `Location` header, so clients will also use the proxy for redirects.
    if (preg_match('/^Location:/', $response_header)) {
        list($header, $value) = preg_split('/: /', $response_header, 2);
        $response_header = 'Location: ' . $_SERVER['REQUEST_URI'] . '?csurl=' . $value;
    }
    if (!preg_match('/^(Transfer-Encoding):/', $response_header)) {
        header($response_header, false);
    }
}

// finally, output the content
print($response_content);

function csajax_debug_message($message)
{
    if (true == CSAJAX_DEBUG) {
        print $message . PHP_EOL;
    }
}


diy_sun.js
'use strict';

var SunCalc = require('suncalc');
var moment = require('moment');
var request = require('sync-request');
var fastXmlParser = require('fast-xml-parser');


function calculateHourWh(panels, lat, lon, sunrise, sunset, sun, hour) {

    var sum = 0;
    var count = 0;
    for (var min = 0; min < 60; min++) {
        hour.set('minute',min);
        hour.set('second',59);
        if (hour.isAfter(sunrise) && hour.isBefore(sunset)) {
            var sunPos = SunCalc.getPosition(hour, lat, lon);
            var sunAzimuth = sunPos.azimuth * 180 / Math.PI;
            var sunAngle = 90 - sunPos.altitude * 180 / Math.PI;

            var panelsWh = 0;
            panels.forEach(function(panel) {
                var azimuthDiff = Math.abs(panel.azimuth - sunAzimuth);
                if (azimuthDiff > 90 ) {
                    azimuthDiff = 90;
                }
                var azimuthFactor = 1 - azimuthDiff/90
                var angleDiff = Math.abs(panel.angle - sunAngle);
                if (angleDiff > 90 ) {
                    angleDiff = 90;
                }
                var angleFactor = 1 - angleDiff/90
                var wh = panel.wattPeak * sun * azimuthFactor * angleFactor;
                panelsWh +=wh;
            });
            sum +=panelsWh;
            count++;
        }
    }
    var hourWh = 0;
    if (count>0) {
        hourWh = sum / count;
    }
    return hourWh;
}

function getYR(panels_cfg, lat, lon, msl, sunrise, sunset) {
    var res = request('GET', 'https://api.met.no/weatherapi/locationforecast/1.9/?lat='+lat+'&lon='+lon+'&msl='+msl);
    var options = {
        attributeNamePrefix : "",
        attrNodeName: false,
        textNodeName : "#text",
        ignoreAttributes : false,
        ignoreNameSpace : false,
        allowBooleanAttributes : false,
        parseNodeValue : true,
        parseAttributeValue : true,
        trimValues: true,
        decodeHTMLchar: false,
    };
    var yr = fastXmlParser.parse(res.getBody('utf8'),options);

    var totalWh = 0;

    var sunriseHour = sunrise.clone().set('minute',0).add(-1,'minute');

    yr.weatherdata.product.time.forEach(function(time){
        var hour = moment(time.from);
        if (hour.isSameOrAfter(sunriseHour) && hour.isBefore(sunset) && time.location.cloudiness) {
            var low = time.location.lowClouds.percent;
            var medium = time.location.mediumClouds.percent;
            var high = time.location.highClouds.percent;

            var clouds = (parseFloat(low)*0.9 + parseFloat(medium)*0.5 + parseFloat(high)*0.3) / 100;
            if (clouds>1.0) {
                clouds = 1.0;
            }
            var sunniness = 1.0 - clouds;
            if (sunniness < 0.15) {
                sunniness = 0.15;
            }
            var hourWh = calculateHourWh(panels_cfg, lat, lon, sunrise, sunset, sunniness,hour);
            totalWh += hourWh;
        }
    });
    return totalWh;
}

function getDarksky(panels_cfg, lat, lon, darkskyApi, sunrise, sunset) {
    var res = request('GET', 'https://api.darksky.net/forecast/'+darkskyApi+'/'+lat+','+lon+"?exclude=daily,currently,minutely,alerts,flags");
    var json = JSON.parse(res.getBody('utf8'));

    var totalWh = 0;

    var sunriseHour = sunrise.clone().set('minute',0).add(-1,'minute');

    json.hourly.data.forEach(function(time){
        var hour = moment(time.time*1000);
        if (hour.isSameOrAfter(sunriseHour) && hour.isBefore(sunset) && time.cloudCover) {

            var clouds = time.cloudCover;

            var sunniness = 1.0 - clouds;
            if (sunniness < 0.15) {
                sunniness = 0.15;
            }

            var hourWh = calculateHourWh(panels_cfg, lat, lon, sunrise, sunset, sunniness,hour);
            totalWh += hourWh;
        }
    });
    return totalWh;
}

/*
example for
panels_cfg = [
        // azimuth: in degrees (direction along the horizon, measured from south to west), e.g. 0 is south and 90 is west
        // angle: angle the panels are facing relative to the horizon in degrees, e.g. 0 they are vertical, 90 they are flat
        // wattPeak: the Wp of the panels
        {
            name: 'Roof',
            azimuth: 8,
            angle: 45,
            wattPeak: 1560
         },
         {
            name: 'Terrace',
            azimuth: 8,
            angle: 57,
            wattPeak: 1560
         }
        ];
*/

exports.solar_prediction_kwh = function(panels_cfg, day_offset, lat, lon, msl, darkskyApi, yr_weight, darksky_weight) {

    var day = moment();

    //console.log(new Date());

    day = moment().add(day_offset, 'day');
    //console.log(day);

    var times = SunCalc.getTimes(day, lat,lon);
    //console.log(times);
    var sunrise = moment(times.sunrise);
    var sunset = moment(times.sunset);

//    console.log(sunrise.format('YYYY-MM-DD')+ " sunrise: "+sunrise.format('HH:mm')+" sunset: "+sunset.format('HH:mm'));

    var yr = getYR(panels_cfg, lat, lon, msl, sunrise,sunset);

    if (typeof darkskyApi !== 'undefined' && darkskyApi !='') {
        var darksky = getDarksky(panels_cfg, lat, lon, darkskyApi, sunrise,sunset);
        var prediction = Math.round(yr * yr_weight + darksky * darksky_weight);
  //      console.log('yr:'+Math.round(yr)+' darksky:'+Math.round(darksky)+' ~= '+ prediction);

    } else {
        prediction = yr;
    }

    return prediction  / 1000;
};



static/logic.js

// ---- configuration ------
var energy = "/energy";
var soc = "/soc";

var max_inactives_seconds = 180;

// feed names
var solar_kwh_name = "solar kwh";
var grid_kwh_name = "grid kwh";
var house_kwh_name = "house kwh";
var powerwall_soc_name = "powerwall soc";

//---------------------------------------------

var solar_watt = 0;
var solar_kwh = 0;
var grid_watt = 0;
var grid_kwh = 0;
var house_watt = 0;
var house_kwh = 0;
var powerwall_watt = 0;
var powerwall_soc = 0;
var solar_prediction = 0;

function loadValues(url) {
    $.getJSON(url, function(data){
        updateValues(data);
        refresh_ui();
    });
}

function updateValues(data) {
    var now = Math.round(new Date().getTime()/1000);

    for (var i = 0, len = data.length; i < len; i++) {
       // SOLAR
//       if (data[i].name == solar_watt_name) {
//         if(data[i].time + max_inactive_seconds >= now){
//            solar_watt = data[i].value
//         } else {
//            solar_watt = 0;
//         }
//       }
       if (data[i].name == solar_kwh_name) {
         solar_kwh = data[i].value;
       }
        // GRID
//       if (data[i].name == grid_watt_name) {
//         if(data[i].time + max_inactive_seconds >= now){
//             grid_watt = data[i].value;
//         } else {
//            grid_watt = 0;
//         }
//       }
       if (data[i].name == grid_kwh_name) {
         grid_kwh = data[i].value;
       }

        // HOUSE
//       if (data[i].name == house_watt_name) {
//         if(data[i].time + max_inactive_seconds >= now){
//             house_watt = data[i].value;
//         } else {
//            house_watt = 0;
//         }
//       }
       if (data[i].name == house_kwh_name) {
         house_kwh = data[i].value;
       }
        // POWERWALL
//       if (data[i].name == powerwall_watt_name) {
//         if(data[i].time + max_inactive_seconds >= now){
//            powerwall_watt = data[i].value;
//         } else {
//            powerwall_watt = 0;
//         }
//       }
       if (data[i].name == powerwall_soc_name) {
         powerwall_soc = data[i].value;
         if (powerwall_soc>100) {
           powerwall_soc=100;
        }
       }
    }
}

function electronTime(watt) {
    w = Math.abs(watt);
    if (w<50) {
        return 6;
    }
    if (w<200) {
        return 4;
    }
    if (w<500) {
        return 2
    }
    if (w<1000) {
        return 1;
    }
    if (w<2000) {
        return 0.5;
    }
    if (w < 3000) {
       return 0.25;
    }
    return 0.125;
}

function setAnimationTime(watt,selector1,selector2) {
    var time = electronTime(watt)+"s";

    $(selector1).each(function( index ) {
        if ($(this).attr("dur")!=time) {
            $(this).attr("dur",time);
        }
    });
   $(selector2).each(function( index ) {
        if ($(this).attr("begin")!=time) {
            $(this).attr("begin",time);
            $(this).attr("dur",time);
        }
    });
}

function refresh_ui() {
    if (solar_watt>0) {
        $("#solar").addClass("on");
        $("#solar_watt").text(solar_watt+" w");
        $("#solar-line").addClass("on");
        $("#solar-dot").addClass("on");
    } else {
        $("#solar").removeClass("on");
        $("#solar-line").removeClass("on");
        $("#solar-dot").removeClass("on");
    }
    prec = (solar_kwh>0)? 2:1;
    $("#solar_kwh").text(parseFloat(solar_kwh).toPrecision(prec)+" kWh");
    setAnimationTime(solar_watt,"#solar-dot animate.dot1","#solar-dot animate.dot2");

    if (grid_watt!=0) {
        $("#grid").addClass("on");
        $("#grid_watt").text(Math.round(grid_watt)+" w");
        $("#grid-line").addClass("on");
        if (grid_watt>0) {
            $("#grid-dot-out").addClass("on");
            $("#grid-dot-in").removeClass("on");
        } else {
            $("#grid-dot-in").addClass("on");
            $("#grid-dot-out").removeClass("on");
        }
    } else {
        $("#grid").removeClass("on");
        $("#grid-line").removeClass("on");
        $("#grid-dot-out").removeClass("on");
        $("#grid-dot-in").removeClass("on");
    }
    prec = (grid_kwh>0)? 1:1;
    $("#grid_kwh").text(parseFloat(grid_kwh).toPrecision(prec)+" kWh");
    setAnimationTime(grid_watt,"#grid-dot-in animate.dot1, #grid-dot-out animate.dot1","#grid-dot-in animate.dot2, #grid-dot-out animate.dot2");

    if (house_watt>0) {
        $("#house").addClass("on");
        $("#house_watt").text(Math.round(house_watt)+" w");
        $("#house-line").addClass("on");
        $("#house-dot").addClass("on");
    } else {
        $("#house").removeClass("on");
        $("#house-line").removeClass("on");
        $("#house-dot").removeClass("on");
    }
    prec = (house_kwh>0)? 2:1;
    $("#house_kwh").text(parseFloat(house_kwh).toPrecision(prec)+" kWh");

    setAnimationTime(house_watt,"#house-dot animate.dot1, #house animate.glow","#house-dot animate.dot2");

    if (Math.abs(powerwall_watt)>2) {
        $("#powerwall").addClass("on");
        $("#powerwall_watt").text(Math.round(powerwall_watt)+" w");
        $("#powerwall-line").addClass("on");
        if (powerwall_watt>0) {
            $("#powerwall-dot-out").removeClass("on");
            $("#powerwall-dot-in").addClass("on");
        } else {
            $("#powerwall-dot-out").addClass("on");
            $("#powerwall-dot-in").removeClass("on");
        }
    } else {
        $("#powerwall").removeClass("on");
        $("#powerwall-line").removeClass("on");
        $("#powerwall-dot-out").removeClass("on");
        $("#powerwall-dot-in").removeClass("on");
    }
    $("#powerwall_soc").text(powerwall_soc+"%");

    setAnimationTime(powerwall_watt,"#powerwall-dot-in animate.dot1, #powerwall-dot-out animate.dot1, #powerwall animate.glow","#powerwall-dot-in animate.dot2, #powerwall-dot-out animate.dot2");

}

loadValues(energy);
loadValues(soc);

var socket = io.connect('/');

socket.on('house', function(data) {
    house_watt = parseInt(data.message);
    refresh_ui();
});

socket.on('powerwall', function(data) {
    powerwall_watt = parseInt(data.message);
    refresh_ui();
});

socket.on('solar', function(data) {
    solar_watt = parseInt(data.message);
    refresh_ui();
});

socket.on("solar prediction", function(data){
    $("#solar_prediction").text(data.message);
});

var interval = setInterval(function() {
    loadValues(energy);
    loadValues(soc);
}, 60000);



project.js
//*********************
// BASIC configuration
const mqtt_host = "mqtt";
const influx_host = "localhost";

//**********************************************
// Solar prediction configuration
// for this to work, you must enter correct values for your location and panel configuration
const latitude = 52.0;
const longitude = 20.0;
const msl = 115; // Mean sea level (in meters)
//
const panels_cfg = [
        // azimuth: in degrees (direction along the horizon, measured from south to west), e.g. 0 is south and 90 is west
        // angle: angle the panels are facing relative to the horizon in degrees, e.g. 0 they are vertical, 90 they are flat
        // wattPeak: the Wp of the panels
        {
            name: 'Roof',
            azimuth: 8,
            angle: 45,
            wattPeak: 1560
         },
         {
            name: 'Terrace',
            azimuth: 8,
            angle: 57,
            wattPeak: 1560
         }
        ];
// API key for darksky.net, in order to use aditionally forecasr from darksky.net you need to get and enter an API key
const darkskyApi = '';

// the final calculated prediction will use weighted average from yr and darksky
// these are the weights for each forecast service, the sum should be 1.0
// adjust to fit which service better forecasts your location
//
const yr_weight = 0.6;
const darsksky_weight = 0.4;


//**************************************************
// Influx kWh queries
const timezone = "Europe/Warsaw";
const house_kwh_query = "SELECT sum(kwh) as kwh FROM energy WHERE (device = 'house' OR device= 'house2') AND time > now() - 1d GROUP BY time(1d) fill(0) TZ('"+timezone+"')";
const grid_kwh_query = "SELECT sum(kwh) as kwh FROM energy WHERE (device = 'sdm120') AND time > now() - 1d GROUP BY time(1d) fill(0) TZ('"+timezone+"')"
const solar_kwh_query = "select sum(wh)/1000 as kwh from (select watt/60 as wh from(SELECT MEAN(charging_power) as watt FROM pcm_query_general_status WHERE time > now() - 1d GROUP BY time(60s) fill(null) TZ('"+timezone+"') )) group by time(1d) fill(0) TZ('"+timezone+"')";
const powerwall_soc_query = "SELECT last(ShuntSOC) FROM generic";

//**************************************************
// MQTT watt topics
const house_watt_topic = 'house/watt';
const house2_watt_topic = 'house2/watt';
const grid_watt_topic = 'grid/watt';
const powerwall_watt_topic = 'powerwall/watt';
const solar_watt_topic = "solar/watt";


//**************************************************
//**************************************************

const express = require('express')
const app = express()

const diy_sun = require('./diy_sun');

const mqtt = require('mqtt')
const mqtt_client = mqtt.connect('mqtt://'+mqtt_host);

const Influx = require('influx')
const influx = new Influx.InfluxDB({
  host: influx_host,
  database: 'powerwall'
})

const influx_batrium = new Influx.InfluxDB({
  host: influx_host,
  database: 'batrium'
})


var moment = require('moment');
var schedule = require('node-schedule');


var house = 0;
var house2 = 0;
var grid = 0;
var powerwall = 0;
var solar = 0;
var solar_prediction = 0;
var solar_prediction_dayOffset = 0;

var server = require('http').createServer(app);
var io = require('socket.io')(server);
var SunCalc = require('suncalc');

function emitSolarPrediction() {
    prefix = "";

    if (solar_prediction_dayOffset>0) {
        prefix = "tomorrow ";
    }

    io.emit('solar prediction', { message: "("+prefix+"prediction "+solar_prediction.toPrecision(2)+" kWh)" });
}

function calculateSolarPrediction() {
    return diy_sun.solar_prediction_kwh( panels_cfg, solar_prediction_dayOffset, latitude, longitude, msl, darkskyApi, yr_weight, darsksky_weight);
}

// schedule every night to recalculate solar prediction for current day and schedule a job for the sunset

var j1 = schedule.scheduleJob('0 2 * * *', function(){

  solar_prediction_dayOffset = 0;
  solar_prediction = calculateSolarPrediction();
  emitSolarPrediction();

  var j2 = schedule.scheduleJob(moment(SunCalc.getTimes(moment(), latitude,longitude).sunset).toDate(), function(){
      // at sunset recalculate solar prediction for tomorrow
      solar_prediction_dayOffset = 1;
      solar_prediction = calculateSolarPrediction();
      emitSolarPrediction();
  });

});


console.log("Calculating solar prediction");
solar_prediction_dayOffset = moment().isAfter(moment(SunCalc.getTimes(moment(), latitude,longitude).sunset))? 1 : 0;
solar_prediction = calculateSolarPrediction();

io.on('connection', function(){
//    console.log("NEW CONNECTION sending all values on start");
    if (house > 0 || house2 > 0) {
        io.emit('house', { message: house+house2 });
    }
    if (powerwall != 0) {
        io.emit('powerwall', { message: powerwall });
    }
    if (grid != 0) {
        io.emit('grid', { message: grid });
    }
    if (solar > 0) {
        io.emit('solar', { message: solar });
    }
    if (solar_prediction > 0) {
        emitSolarPrediction();
    }
});

mqtt_client.on('connect', () => {
//  console.log("MQTT connected");
  mqtt_client.subscribe(house_watt_topic);
  mqtt_client.subscribe(house2_watt_topic);
  mqtt_client.subscribe(grid_watt_topic);
  mqtt_client.subscribe(powerwall_watt_topic);
  mqtt_client.subscribe(solar_watt_topic);
})

mqtt_client.on("close",function(error){
   console.log("mqtt can't connect"+error);
   io.emit('house', { message: 0 });
   io.emit('grid', { message: 0 });
   io.emit('powerwall', { message: 0 });
   io.emit('solar', { message: 0 });
 })

mqtt_client.on('message', (topic, message) => {
  if(topic === house_watt_topic) {
       house = parseInt(message.toString())
       io.emit('house', { message: house+house2 });
  } else
  if(topic === powerwall_watt_topic) {
       powerwall = parseInt(message.toString())
       io.emit('powerwall', { message: powerwall });
  } else
  if(topic === grid_watt_topic) {
       grid = parseInt(message.toString())
       io.emit('grid', { message: grid });
  } else
  if(topic === house2_watt_topic) {
       house2 = parseInt(message.toString())
        io.emit('house', { message: house+house2 });
  } else
  if(topic === solar_watt_topic) {
       solar = parseInt(message.toString())
       io.emit('solar', { message: solar });
   }
})

app.get('/energy', function (req, res) {
  influx.query(house_kwh_query).then( house => {
    influx.query(grid_kwh_query).then ( grid => {
        influx.query(solar_kwh_query).then ( solar => {
            res.json(
                [
                    {"name":"grid kwh","value": (grid === undefined || grid.length == 0)? 0 : grid[1].kwh},
                    {"name":"house kwh","value":house[1].kwh},
                    {"name":"solar kwh","value":solar[solar.length -1].kwh}
                ]
                );
        });
    });
  });
})

app.get('/soc', function (req, res) {
  influx_batrium.query(powerwall_soc_query).then( soc => {
    res.json(
        [
            {"name":"powerwall soc","value":soc[0].last}
        ]
        );
  });
})


app.use(express.static('static'))

server.listen(3333, () => console.log('DIY powerflow listening on port 3333!'))

